Suggestions for parameter hierarchy (#150)
openclimatedata/openscm: docs/parameters.rst

.. _standard-parameters:

Standard parameters
===================

> In OpenSCM a ':ref:`parameter <parameter-hierarchy>`' is any named input or output variable of a mode, e.g. |CO2| emissions, equilibrium climate sensitivity, aerosol forcing scaling.

Reviewer: ```suggestion
In OpenSCM a ':ref:`parameter <parameter-hierarchy>`' is any named input or output variable of a model, e.g. |CO2| emissions, equilibrium climate sensitivity, aerosol forcing scaling.
```

As described :ref:`here <parameter-hierarchy>`, parameters are given in a hierarchy, e.g. ``Emissions`` -> ``CO2`` -> ``Industrial``.

Simple climate models come in many different shapes and forms hence we do not expect them to all be able to do everything.

> However, to be included in OpenSCM they should make sure their parameters fit into these standard parameters as far as possible to ensure models can be interchanged easily.

Reviewer: I don't like this change. They have to understand all of the key parameters, where 'understand' means that their adapters should either be able to use them or should throw sensible warning/error messages if they are used. If they just blow up when given a key parameter then they're useless.

Author: Determining which parametres were used should not be task of an adapter (see #123).

> If they just blow up when given a key parameter then they're useless.

Why should they?

Reviewer: We're not on the same page. 

What I want to make clear is that the user should always be able to pass key parameters e.g. `ECS` or `TCR` to a high-level `run` call and get sensible output i.e. things either run, a clear warning is raised or a clear error is raised. We don't want users passing 'standard' parameters in, only to receive a cryptic error message (which is possible if the implementer of an adapter doesn't think about how all (probably 4 or 5) of the key parameters would be handled). 

The text you've suggested does not make this clear ('as far as possible' is vague and doesn't tell the implementer what they actually have to do). The text I have clearly also fails to communicate what I wanted so I think we need to re-write this sentence somehow.

Doing this checking has to be done somewhere. I thought the most natural place was an adapter but maybe I'm missing something?

Author: > We're not on the same page.

No ;)

> What I want to make clear is that the user should always be able to pass key parameters e.g. `ECS` or `TCR` to a high-level `run` call and get sensible output i.e. things either run, a clear warning is raised or a clear error is raised. We don't want users passing 'standard' parameters in, only to receive a cryptic error message (which is possible if the implementer of an adapter doesn't think about how all (probably 4 or 5) of the key parameters would be handled).

I don't understand why that would lead to errors at all if parameters are simply ignored by the adapter because it does not know about them. I think we must not make all adapters use every parameter described here, many models would not be able to use many of these anyway (e.g. non-carbon-gases for DICE). Adapter developers should only make sure that the parameters for the respective model are consistent with these standard parameters (e.g. giving ECS a different name).

> The text you've suggested does not make this clear ('as far as possible' is vague and doesn't tell the implementer what they actually have to do). The text I have clearly also fails to communicate what I wanted so I think we need to re-write this sentence somehow.

Maybe like the text in m,y previous reply?

> Doing this checking has to be done somewhere. I thought the most natural place was an adapter but maybe I'm missing something?

I would do that in the Core so adapters do not have to be altered when we add or change standard parameters.

Reviewer: Right I get it now. Handling this in core will also work so let's go for something like

"
However, to be included in OpenSCM they should make sure their parameters fit into these standard parameters as far as possible. 
"

We can then work out the details as we start building useful tools around the core <-> adapter connection

Of course, also model-specific parameters are possible (see also :ref:`writing-adapters`).

Conventions
-----------

In the following, 'pre-industrial' refers to an unperturbed state of the climate.
Individual adapters can translate this into whatever year they need to for their model, but they should do such translations with this definition in mind.

'Reference period' refers to the period a given variable is reported relative to the mean of.
For example, 'surface temperature relative to a 1961-1990 reference period' refers to surface temperatures relative to the mean of the period 1961-1990.
Adapters can report variables without reference periods if these are not used internally by the models.
However, if the model uses an internal reference period then this should be indicated in the reported variable names by appending ``(rel. to XXXX-YYYY)`` to the variable name.
Hence, 'surface temperature relative to a 1961-1990 reference period' would become ``Surface Temperature (rel. to 1961-1990)``.
[TODO: discuss with other authors how we want this to play out]

Aggregation
-----------

> Parameters in OpenSCM come as part of a hierarchy, in the following separated by the ``|`` character.

Reviewer: ```suggestion
Parameters in OpenSCM come as part of a hierarchy, separated by the ``|`` character.
```

They're always separated by this character no? Or do you envisage being able to separate different hierarchical separators?

Author: Not in the API...

Reviewer: definitely outdated/unthinking comment by me, oops

For example, ``Emissions|CO2|Energy``.
``Emissions|CO2|Energy`` is emissions of |CO2| from the energy sub-sector (whatever 'energy' happens to mean in this context).
As far as it makes sense, parameters that are higher in the hierarchy (e.g. ``Emissions|CO2`` is 'higher' than ``Emissions|CO2|Energy``) are the sum of all the variables which are one level below them in the hierarchy.

> For example, for ``Emissions|CO2|Energy``, ``Emissions|CO2|Transport`` and ``Emissions|CO2|Agriculture`` given, ``Emissions|CO2`` would be the sum of these.

Reviewer: ```suggestion
For example, if ``Emissions|CO2|Energy``, ``Emissions|CO2|Transport`` and ``Emissions|CO2|Agriculture`` are provided, ``Emissions|CO2`` would be the sum of these.
```

Standards
---------

Standard parameters
*******************

> Below we provide a list of the OpenSCM standard parameters adapters must adhere to as far as a specific variable concerns them.

Reviewer: ```suggestion
Below we provide a list of the OpenSCM standard parameters adapters must understand to as far as a specific variable concerns them.
```

Alongside we give the type of unit that the parameter should be given in and how it should be expected by adapters.
Conversion between particular units is done automatically if possible.

In the following, ``<GAS>`` can be one of the standard :ref:`gases`.

.. csv-table:: Standard parameters
    :header: "Parameter name 0", "Parameter name 1", "Parameter name 2", "Unit type", "Note"

    ``Emissions``, ``<GAS>``,, "mass <GAS> / time"

>     ``Atmospheric Concentrations``, ``<GAS>``,, "concentration (e.g. parts per X)", "Aggregation possible, but does not always make sense"

Reviewer: ```suggestion
    ``Atmospheric Concentrations``, ``<GAS>``,, "parts per X where X is million, billion, trillion etc.", "Aggregation possible, but does not always make sense"
```

>     ``Pool``, ``<GAS>``,, "mass <GAS>"

Reviewer: I wonder if we can do better than 'Pool', maybe not...

Author: Hmm, it's the standard expression, no? "Gas pool"?

Reviewer: I think so, it just seems unsatisfactory that it is so...

    ``Radiative Forcing``, ``Aerosols``
    ``Radiative Forcing``, ``Aerosols``, ``Direct Effect``
    ``Radiative Forcing``, ``Aerosols``, ``Indirect Effect``
    ``Radiative Forcing``, ``Aerosols``, ``SOx``
    ``Radiative Forcing``, ``Aerosols``, ``NOx``
    ``Radiative Forcing``, ``Aerosols``, ``OC``
    ``Radiative Forcing``, ``Aerosols``, ``BC``
    ``Radiative Forcing``, ``Land-use Change``
    ``Radiative Forcing``, ``Black Carbon on Snow``
    ``Radiative Forcing``, ``Volcanic``
    ``Radiative Forcing``, ``Solar``
    ``Radiative Forcing``, ``External``
    ``X to Y Flux``,,, "mass / time", "See :ref:`material_fluxes`"
    ``Surface Temperature``,,, "temperature", "Surface air temperature i.e. ``tas``"
    ``Ocean Temperature``,,, "temperature", "Surface ocean temperature i.e. ``tos``"
    ``Ocean Heat Content``,,, "energy"
    ``Sea Level Rise``,,, "length"
    ``Equilibrium Climate Sensitivity``,,, "temperature"
    ``Transient Climate Response``,,, "temperature"
    ``f2xco2``,,, "power / area", "Radiative forcing due to a doubling of atmospheric |CO2| concentrations from pre-industrial level"

.. _gases:

Gases
*****

.. csv-table:: Gases

>     :header: "Name", "Description"

Reviewer: do we want an 'also known as' column here? For example 'HFC365mfc' is the same as 'HFC365-mfc'. It could take inspiration from @rgieseke's greenhouse gas renamer project

Author: Hmm, could be part of the "Description" if there are really non-trivial variants... Probably we should better properly describe HFC365mfc et al...

Reviewer: ok I'll add in once this is merged

.. _material_fluxes:

Material Fluxes
***************

These variables can be used to store the flux of material within the model.
They should be of the form ``X to Y Flux`` where the material is flowing from ``X`` into ``Y`` (and hence negative values represent flows from ``Y`` into ``X``):

- ``Land to Air Flux|CO2|Permafrost`` (mass carbon / time) - land to air flux of |CO2| from permafrost
- ``Land to Air Flux|CH4|Permafrost`` (mass methane / time)

Standard regions
----------------

Similarly to variables, regions are also given in a hierarchy.
Regions which are higher in the hierarchy are the sum of all the regions which are one level below them in the hierarchy (be careful of this when looking at e.g. |CO2| concentration data at a regional level).

.. csv-table:: Gases
    :header: "Name 0", "Name 1", "Name 2, "Description"

    ``World``
    ``World``, ``Northern Hemisphere``
    ``World``, ``Northern Hemisphere``, ``Ocean``
    ``World``, ``Northern Hemisphere``, ``Land``
    ``World``, ``Southern Hemisphere``
    ``World``, ``Southern Hemisphere``, ``Ocean``
    ``World``, ``Southern Hemisphere``, ``Land``
    ``World``, ``Ocean``
    ``World``, ``Land``

>     ``World``, ``R5ASIA``

Reviewer: can you add a TODO for me here to dig out the links of where all these come from?

Reviewer: @swillner ping

Author: Sorry ;) Done
